src/realm/object-store/results.hpp

```diff
@@ -335,20 +336,50 @@ class Results {
         m_update_policy = policy;
     }
 
+    /**
+     * Creates a SectionedResults object by using a user defined sectioning algorithm to project the key for each
+     * section.
+     *
+     * @param section_key_func The callback to be itterated on each value in the underlying Results.
+     * This callback must return a value which defines the section key
+     *
+     * @return A SectionedResults object using a user defined sectioning algoritm.
+     */
+    SectionedResults
+    sectioned_results(util::UniqueFunction<Mixed(Mixed value, std::shared_ptr<Realm> realm)> section_key_func);
+    enum class SectionedResultsOperator {
+        FirstLetter // Section by the first letter of each string element. Note that col must be a string.
+    };
+
+    /**
+     * Creates a SectionedResults object by using a built in sectioning algorithm to help with efficiency and reduce
+     * overhead from the SDK level.
+     *
+     * @param op The `SectionedResultsOperator` operator to use
+     * @param property_name Takes a property name if sectioning on a collection of links, the property name needs to
+     * reference the column being sectioned on.
+     *
+     * @return A SectionedResults object with results sectioned based on the chosen built in operator.
+     */
+    SectionedResults sectioned_results(SectionedResultsOperator op,
+                                       util::Optional<StringData> property_name = util::none);
+
 private:
     std::shared_ptr<Realm> m_realm;
     mutable util::CopyableAtomic<const ObjectSchema*> m_object_schema = nullptr;
     Query m_query GUARDED_BY(m_mutex);
-    TableView m_table_view GUARDED_BY(m_mutex);
     ConstTableRef m_table;
+    TableView m_table_view GUARDED_BY(m_mutex);
     DescriptorOrdering m_descriptor_ordering;
     std::shared_ptr<CollectionBase> m_collection;
     util::Optional<std::vector<size_t>> m_list_indices GUARDED_BY(m_mutex);
 
     _impl::CollectionNotifier::Handle<_impl::ResultsNotifierBase> m_notifier;
 
     Mode m_mode GUARDED_BY(m_mutex) = Mode::Empty;
+    friend class SectionedResults;
     UpdatePolicy m_update_policy = UpdatePolicy::Auto;
+    uint64_t m_last_collection_content_version GUARDED_BY(m_mutex);
```

Comment:
We need to keep track of the current `m_collection` content version as a new instance of `m_coll_base` is created every time on property access due to how Cocoa's property accessors work, meaning that tracking the content version inside `CollectionBase` can produce stale results if we try to hold onto an instance of it else where.